lib/ASRouterTriggerListeners.jsm

```diff
@@ -65,27 +78,37 @@ this.ASRouterTriggerListeners = new Map([
     _initialized: false,
     _triggerHandler: null,
     _hosts: null,
+    _matchPatternSet: null,
     _visits: null,
 
-    async init(triggerHandler, hosts) {
-      if (this._initialized) {
-        return;
-      }
-      this.onTabSwitch = this.onTabSwitch.bind(this);
+    async init(triggerHandler, hosts = [], patterns) {
+      if (!this._initialized) {
+        this.onTabSwitch = this.onTabSwitch.bind(this);
 
-      // Add listeners to all existing browser windows
-      for (let win of Services.wm.getEnumerator("navigator:browser")) {
-        if (isPrivateWindow(win)) {
-          continue;
+        // Add listeners to all existing browser windows
+        for (let win of Services.wm.getEnumerator("navigator:browser")) {
+          if (isPrivateWindow(win)) {
+            continue;
+          }
+          await checkStartupFinished(win);
+          win.addEventListener("TabSelect", this.onTabSwitch);
+          win.gBrowser.addTabsProgressListener(this);
         }
-        await checkStartupFinished(win);
-        win.addEventListener("TabSelect", this.onTabSwitch);
-        win.gBrowser.addTabsProgressListener(this);
-      }
 
-      this._initialized = true;
+        this._initialized = true;
+        this._visits = new Map();
+      }
       this._triggerHandler = triggerHandler;
-      this._visits = new Map();
+      if (patterns) {
+        if (this._matchPatternSet) {
+          this._matchPatternSet = new MatchPatternSet([
+            ...this._matchPatternSet.patterns,
+            ...patterns,
```

Comment:
It might be worth removing duplicates here (for example, if we have a bunch of messages that target the same site, we would add in multiple copies)